### ISSUE TITLE
Utility object for iterating bytestrings

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6043]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6043|
|Type|enhancement|
|Created|2012-10-03 12:33:46Z|
|Branch|https://github.com/twisted/twisted/tree/bytes-py3-6043|

Python 3 does not provide a way to iterate over individual bytes in a bytestring. We should provide a cross-version utility wrapper in `twisted.python.compat`.

A future enhancement might add `%` support on Python 3, another would fix indexing (e.g. `b'abc[1]'`) to not return an integer.

<details><summary>Searchable metadata</summary>

```
trac-id__6043 6043
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_bytes_py3_6043 branches/bytes-py3-6043
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1349267626000000 1349267626000000
changetime__1349284970000000 1349284970000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [35905]) Branching to 'bytes-py3-6043'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set owner to @exarkun|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/bytes-py3-6043

I would appreciate suggestions for a better name.

I left news file as a `.misc` on the theory that's only relevant to developers of Twisted, who would be better off reading the Python 3 page as a whole rather than trying to get individual info from news snippets. It could be changed to `feature` though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.

  1. misc seems fine to me
  1. I was expecting something like an `iterbytes` function, rather than a class-or-no-op-function.  Maybe that helps with the naming issue?  Another issue that helps avoid is making the type of the helper depend on Python version.  Not as though anyone should ever try to subclass this thing, but we have no real way to stop them from doing so.  Also the variable type makes it impossible to satisfy Twisted's naming convention with respect to casing.  On the other hand, `iterbytes` doesn't sound like something that would ever help with interpolation (but perhaps we should just have a separate API for interpolation).
  1. This is a bit hair-splitty, but `list([b for b in python2Bytes(input)])` is always equivalent to `list(python2Bytes(input))` right?

Merge when you're happy with it all.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [35915]) Merge bytes-py3-6043: Utility function for iterating over bytes.

Author: itamar
Review: exarkun
Fixes: [#6043](https://github.com/twisted/twisted/issues/6043)

Python 3 bytes are broken. This provides a utility function to deal with one issue out of many.
